Python unittest enhancements and source documentation

ceeli opened the pull request:
This pull request includes changes concerning all of FreeCAD (inside the Mod/Test module) and changes that concern mainly the fem module. See the source/file documentation for more info on the new modules.

There are new test support modules for App tests (AppTestSupport) and Gui tests (GuiTestSupport) in the Mod/Test module. This should be useful for all of FreeCADs python unit tests.

Test modules are now in sub packages of the femtest package. femtest.app contains all test modules which don't need a gui to run. femtest.gui contains all tests that require a gui to run. The test files are in femtest/data. The packages use the load_tests protocol so they can be used as targets when executing tests (e.g. FreeCADCmd -t femtest). This could replace the TestFem module. The packages automatically find all test (they don't need to be listed again).

Test coverage of the python source can now be measured by using the --cov option in addition to the -t option. See the documentation of src/App/FreeCADTest for more detail. coverage.py must be installed.

I don't expect that this pull request is just merged without some tweaking. I'll be available for at least a month and I intend to work on testing and documentation of FreeCAD for much longer than that. This branch doesn't work properly with python2. Is that still a requirement?

Commit:
Add coverage support to FreeCADTest.py script

To get coverage information use the new --cov options in addition to the
-t option when executing tests. Optionally specify a package or module
which shall be measured to avoid overloaded coverage reports
(--cov [<module>]). A coverage summery is produced after the test output
and a html coverage report is generated in the coverage/ subdirectory of
the build directory.

Commit:
FEM: Make use of new test/coverage system

Use test package to make the fem test suite instead of the TestFem.py
module. The test packages also specify coverage information to
automatically generate meaningful coverage reports.

Commit:
Add app test helper module: Test/AppTestSupport.py

The support module provides base classes for FreeCAD unittest TestCase classes.
It does stuff like providing FreeCAD specific asserts and automatic document
management.

Commit:
FEM: Add tests for the femtools/femutils.py module

Commit:
FEM: Add helper for textbook solver unit tests

This module contains base classes for tests that execute whole
simulations and verify (assert) the results. It automates the whole
process from analysis creation to asserting the results as much as
possible.

Commit:
FEM: Add tests for textbook example verification

The added tests execute a predefined simulation and than check the
results against calculated values of a analytical approximation. Makes
heavy use of the test support library. Predifined cases are stored in
normal .FCStd files. Each unittests that adds a solver, configures the
settings and executes the simulation. Most of the work is done by
the support library though.

Commit:
Add gui test helper module: Test/GuiTestSupport.py

The support module provides base classes for FreeCAD unittest TestCase classes.
There will be base classes for each gui element of FreeCAD that can be tested
individually. At the moment there is only one for Task Panels which makes it
very easy to test Task Panels (click stuff, assert graphical output, ...).

Commit:
FEM: Add gmsh task panel tests (first gui tests)

Commit:
Add --log option to set loglevel for one session

The new --log option can be used for the gui and command line binary.
Set it to "Log", "Msg", "Wrn" or "Err" to adjust the loglevel for one
execution.

Commit:
Make Sequencer output sensible to loglevel

The sequences previously printed directly to stdout without checking the
loglevel. Now it prints with the priority "Msg".

Commit:
Use "Err" loglevel default for python test execution

With the previously introduced --log option the loglevel can be set on launch
for one execution of FreeCAD. Very useful for test execution. The default if no
loglevel is specified is normally "Wrn".

berndhahnebach commented:
> This branch doesn't work properly with python2. Is that still a requirement?

AFAIK for 0.19 yes. Even if we force developer and user to use Python 3

On all other stuff. I will have a look and give some feedback.

berndhahnebach commented:
Since it help s a test module and it is new I do not think we need py2 support. The time we need to get all this into FC py2 might be obsolete.

It will take some time, but we need to get started somdhow ... Lets do a simple one. I saw you moved from testfiles to data directory. Is it posible to extract this to a separete commit? I would just merge this into master.

berndhahnebach commented:
damn lots of merge conflicts ...

berndhahnebach commented:
Would it be possible to split it in two PRs one only FEM related commits? We could these merge really easily. But the not FEM related test commits needs to be disscussed first. It would even be much simpler to disscuss these commits if they would be without all these FEM changes. The FEM changes need only to be reviewed from my side. I even can merge them into master directly. 

It is really a pity I have been overoveroverloaded last few month ... 

there is some cools stuff inside ... 

cheers bernd

berndhahnebach commented:
if you replace a print with a FreeCAD.Console.PrintWhatever you need to add a new line because it is not added like print it does.

BTW: this replacement is one of the last remaining FEM code formating.

Have you noticed https://github.com/FreeCAD/FreeCAD/blob/master/src/Mod/Fem/coding_conventions.md

ceeli commented:
Alright lets do this. First let me check where all those merge conflicts come from. There haven't been any a few weeks ago.

berndhahnebach commented:
mainly my code formmating replacing all ' by "

ceeli commented:
Oh, thats troublesome... I think I have to resolve all that by hand than. Well, I'll be back in an hour or so :P

ceeli commented:
Alright I resolved everything. I couldn't test anything cause my FreeCAD build need to completely recompile which takes ~30min on my notebook.

berndhahnebach commented:
cool I will have a look. I allready saw some commits I would like to cherry-pick. Have you seen https://forum.freecadweb.org/viewtopic.php?t=36682

berndhahnebach commented:
still there are lots of problems in the rebase I solve a few here ... hopefully ... https://github.com/FreeCAD/FreeCAD/pull/2549

ceeli commented:
I know I'm solving them right now. Those are just some merge mistakes.

berndhahnebach commented:
I merged all changes not related to your new test suite. cd8f0e0960d048032b6967edb9e8952541a5eb6c

berndhahnebach commented:
I had to splitt some commits. Hopefully you get it all together again.

berndhahnebach commented:
btw: somehow your github email does not fit with your git email. AFAIK you could even set it to some strange mail just to get you found in github as a contributor

berndhahnebach commented:
How about a rebase Markus?

ceeli commented:
Yes, will do. I'm quite busy atm sorry... I'll do it tomorrow.

berndhahnebach commented:
AFAIKT you should only have changes left in the regard of the new test tool, means new code and only a few code changes. Most of the code changes ATM are changes which change stuff back to how it was in summer. 

cheers bernd

berndhahnebach commented:
ping

ceeli commented:
Still here, just very busy with studying, sorry... I'll going to look though it now

berndhahnebach commented:
I did a rebase. Would you give this a try? https://github.com/berndhahnebach/FreeCAD_bhb/commits/gsoctestpy2

do not forget the 2 at branch end

berndhahnebach commented:
made some code formatting and added some licence header.

berndhahnebach commented:
something is broken in the old FEM tests! How do I run the new FEM tests?

Should we move disscussion to the forum?

Commit:
FEM: Restore path functions of run module

The functions getTempDir, getBesideDir and getBesideBase where moved to
femutils, modified and made public in the past. This breaks the solver
framework under certain circumstances. These folders are only intended
for the solver. Using it for other puroses may result in unexpected
behaviour (including crashes). This commit restores the original
functions for the solver module. The problem that other modules of the
workbench use those folders should be resolved at some point.

ceeli commented:
I have a few hours to spare today. Please don't apply anything (from the GSoC Stuff) to femdev otherwise I have to start over again.

berndhahnebach commented:
> 
> 
> I have a few hours to spare today. Please don't apply anything (from the GSoC Stuff) to femdev otherwise I have to start over again.

For sure I will not :-)

I cherry picked your commits. It took me just half an hour. See https://github.com/berndhahnebach/FreeCAD_bhb/commits/gsoctestpy2

Commit:
FEM: Add gui unit test package (femtest/gui)

Analog to the femtest/app package add a femtest/gui package which
contains all tests that depend on the FreeCAD gui. Tests from this
package are only supposed to be executed when the gui is started.

Commit:
FEM: Remove debug print output from test module

There was some output still getting through and makeing the test run
output confusing. These prints where commented out for now.

ceeli commented:
> > I have a few hours to spare today. Please don't apply anything (from the GSoC Stuff) to femdev otherwise I have to start over again.
> 
> For sure I will not :-)
> 
> I cherry picked your commits. It took me just half an hour. See https://github.com/berndhahnebach/FreeCAD_bhb/commits/gsoctestpy2

Thanks! The rebase is now finished and for me everything works. I just pulled the changes from yesterday and today and for some reason the recompile takes quite long now. I have to go to work now but when I come home I'll do one test run and if everything works I'll push it. It took a little longer because I also fixed stuff and cleaned up the history a bit.

ceeli commented:
Now everything should work. The way of executing tests didn't really change, it is now just possible to specify the packages directly e.g.:
`bin/FreeCADCmd -t femtest` or
`bin/FreeCADCmd -t femtest.app`
All the stuff that is not Fem specific is now at the beginning of the commit history the Fem specific stuff is later in the history but largely depends on the general changes.
I would also suggest to adjust this branch till it is ready to be merged instead of cherry-picking everything (tearing it apart) to keep the git overhead low.

ceeli commented:
I wouldn't mind to move the discussion to the forum btw. if you want to do that :)

berndhahnebach commented:
a9c19ca6d42c6bf31eac194c936c215b0dc382f3 and f4e79ed26b5896faa5b1aef7673258c212cc7d9b have been merged into master. The time beeing all your commits not directly related to the new unit tests are in master. 

branch https://github.com/berndhahnebach/FreeCAD_bhb/commits/gsoctestpy has all the rest and will be rebased regularly.

bernd

berndhahnebach commented:
rebased and added pep8 code formating

https://github.com/berndhahnebach/FreeCAD_bhb/commits/gsoctestpy

berndhahnebach commented:
did another rebase and finally had the time to start playing with this ... If run from inside FreeCAD I am able to run the tests. Some needs some fixed becaus there has been some development taken place. But I am not able to run from a bash, at least there is no print message at all. 

https://github.com/berndhahnebach/FreeCAD_bhb/blob/fcb56c38e0/src/Mod/Fem/TestFemApp.py#L359-L407

berndhahnebach commented:
How does the --cov option works? Do you have some example?

Best is to switch to the forum, ahh just saw you posted some exameple some posts ago ...

berndhahnebach commented:
ok it works from bash too :-) 
But how about this --cov option and the html reports? How does this works?

berndhahnebach commented:
Hi Markus
I have been playing a lot with your code. Everything works great so far. The only thing I do not get to work is the coverage report. It just does not create some report for me, nothing. 

commit 22e4dc68da 

`make -j 4 && ./bin/FreeCAD --cov -t femtest
`
Would be coold if you could give some advise in this regard. Find the latest branch here: https://github.com/berndhahnebach/FreeCAD_bhb/commits/gsoctestpy


cheers bernd
